packages/react-scripts/package.json

```diff
@@ -82,7 +82,7 @@
     "webpack": "4.43.0",
     "webpack-dev-server": "3.11.0",
     "webpack-manifest-plugin": "2.2.0",
-    "workbox-webpack-plugin": "4.3.1"
+    "workbox-webpack-plugin": "5.1.3"
```

Comment:
(My knowledge of `yarn` is pretty limited, as I use `npm`.)

As per this [build error](https://dev.azure.com/facebook/create-react-app/_build/results?buildId=1617&view=logs&j=e95a0a11-9168-5bc5-8aec-a080722f0e2b&t=632281fd-9741-5b0c-5338-e74f74038edb&l=10759):

```
Module not found: You cannot require a package ("workbox-core") that is not declared in your dependencies (via "/home/vsts/work/_temp/tmp.uOfOvXy03y/test-app-pnp/src/service-worker.js")
```

I guess I need to add all the Workbox runtime libraries as dependencies here, as developers might use any of them in their service worker.

I tried

```
$ yarn workspace react-scripts add workbox-background-sync workbox-broadcast-update workbox-cacheable-response workbox-core workbox-expiration workbox-google-analytics workbox-navigation-preload workbox-precaching workbox-range-requests workbox-routing workbox-strategies workbox-streams
```

and that led to

```
error An unexpected error occurred: "expected workspace package to exist for \"minimist-options\"".
```

Reply:
The last error i believe is related to a yarn bug. I get the same when using `yarn upgrade-interactive` and applying changes. My workaround has been `npx yarn@1.16.0 upgrade-interactive`. Perhaps that would fix the last error?

I don't believe the proposed solution to add as a dependency in `react-scripts` will solve the issue however. As far as i know [`pnp-webpack-plugin`](https://github.com/arcanis/pnp-webpack-plugin) will make sure that the user's package.json of their project lists the referenced workbox packages or it will error.

@arcanis any suggestions?

Reply:
Oh, that's interesting. FWIW, the `workbox-webpack-plugin` itself depends on `workbox-build`, which in turn depends on all of the individual Workbox packages. So by virtue of having a `workbox-webpack-plugin` dependency, the underlying Workbox packages will actually be installed—but I guess that's not enough to `pnp-webpack-plugin`.

Reply:
I only gave a cursory look, but transitive dependencies aren't valid, @jeffposnick (ie, just because A depends on B, the fact that you depend on A doesn't mean you can access B). Projects such as Babel, ESLint, typically avoid this by loading their plugins on behalf of the configuration file that requires them (using `createRequire` or the `paths` option from `require.resolve`). This ensures that the responsibility of listing the plugin dependency is left to whoever requires it.

Reply:
`npx yarn@1.16.0 workspace react-scripts add ...` worked.

I _think_ that's the right thing to do to resolve the build error, but some of the nuance around what that `createRequire` alternative would entail might be lost on me...

Reply:
...I guess that didn't actually fix the build error, though.

The default `service-worker.js` that is checked in to `cra-template` is, via the Workbox `InjectManifest` plugin, run through a `webpack` child compiler to produce a final compiled `service-worker.js`. `service-worker.js` includes references to `workbox-core` and a few other Workbox libraries, and might end up importing additional ones if the  user ends up making changes to it.

Given that, I'm assuming that there needs to be a reference to at least the `workbox-core` and a few other libraries added _somewhere_  to resolve the build error, since in this case, it's not the end user that's adding in the dependency, but rather its something checked in by default.

Reply:
The way this is done in other all-in-one frameworks is to re-export the module that the transpiled code needs to access in order to go through your own. For example, if the userland code would get transpiled into `import 'foo'`, you'd instead want to transpile it into `import 'react-scripts/foo'`, and have a `foo.js` at the root of your package that would do `module.exports = require('foo')`. This obviously assumes that the transform pass has a way to configure the imported module.

Note that this is a problem for all package managers, not only Yarn - while more complex than expected, if you don't use the pattern I described, you will hit an issue when multiple versions of workbox are in your users' dependency trees (since there's no telling which version will get hoisted to the top).

Reply:
Apologies, but I've read through this and my lack of familiarity with these concepts is preventing me from making progress.

Maybe there's an example of this already checked in to the `c-r-a` repo that could point me in the right direction?

For instance,

https://github.com/facebook/create-react-app/blob/abe236963c75160f905f0cd96a2abaee245ba653/packages/cra-template/template/src/App.js#L1

seems (to me at least) to be equivalent, in that it's an import of the module `react` which is listed as a `devDependency` at

https://github.com/facebook/create-react-app/blob/abe236963c75160f905f0cd96a2abaee245ba653/packages/react-scripts/package.json#L88

I don't see anything special checked in to `c-r-a` to do a re-export of `react` or anything like that.

The calls to

```
import { clientsClaim } from 'workbox-core';
// other workbox-* imports
```

that are in the new `service-worker.js` file, which is processed by a child compiler created by the main `webpack` compilation, _seem_ like they should be equivalent now that I've listed those dependencies explicitly in `react-scripts/package.json`, but there's obviously something different that I don't understand.

I'm not sure how to replicate the [CI test failure](https://dev.azure.com/facebook/create-react-app/_build/results?buildId=1620&view=logs&j=e95a0a11-9168-5bc5-8aec-a080722f0e2b&t=632281fd-9741-5b0c-5338-e74f74038edb&l=10712), as running `yarn build` in my local copy of the repo works fine, and the build completes successfully, so I would basically be submitting commits "blind," hoping to fix an issue in the CI that I can't reproduce locally.